### PR TITLE
Fix `Program<T>` equality

### PR DESCRIPTION
**Pull request by zuqq**
The version on `main` compares `self.instructions` to itself:

https://github.com/Smaug123/performance_aware_programming/blob/00d8fcf98d83c6a0ee2ec41aec9eabef95748fe1/src/main.rs#L570-L580

### CODE REVIEW

**Comment by Smaug123** [comment]
No wonder my tests were passing

File: src/main.rs [hunk 1]
```diff
@@ -681,11 +658,48 @@ fn main() {
...
+                Register::General(GeneralRegister::D, RegisterSubset::All),
+                1,
+            ))],
+        };
+
+        assert_ne!(program1, program2);
+    }
+
+    #[test]
+    fn test_programs_with_identical_instruction_sequences_are_equal() {
```
Comment (Smaug123) [inline src/main.rs:682]: This is where I need to get out the property based testing Rust libraries

**Review by Smaug123 — DISMISSED** [review]

**Comment by Smaug123** [comment]
Oh no, an actual bug!

File: src/main.rs (hunk 1 above)
Comment (zuqq) [reply src/main.rs:682]: This is just me trying to convince myself that the derived implementation is sensible.

**Comment by Smaug123** [comment]
This is a test that currently fails:
```rust
    #[test]
    fn mem_reg_move_sum_to_bytes() {
        let i = Instruction::MemRegMove(MemRegMove {
            source: crate::EffectiveAddress::Sum(crate::WithOffset::WithU16((Base::Bx, SourceDest::Source), 4999)),
            dest: Register::General(GeneralRegister::D, RegisterSubset::All),
        });
        assert_eq!(i.to_bytes(), vec![138, 128, 135, 19]);
    }
```

File: src/main.rs [hunk 2]
```diff
@@ -449,14 +449,14 @@ impl Instruction {
                         } else {
                             Base::Bp
                         };
-                        let displacement_high = if rm == 6 || mode > 0 {
+                        let displacement_low = if rm == 6 || mode > 0 {
                             bytes.next().expect("required an 8-bit displacement")
                         } else {
                             0
                         };
-                        let displacement_low = if (rm == 6 && mode == 0) || mode == 2 {
```
Comment (Smaug123) [inline src/main.rs:457]: Voici the bug.

**Review by Smaug123 — APPROVED** [review]

**Pull request merged**